### PR TITLE
Change Azure keyVault to use azure:AdministratorGroups

### LINKED ISSUE #120: Feature - Baseline Prologue for AzureAdministratorsGroup
## Expected Behavior

The baseline component should be suitable for all new products and tenants out of the box.

## Current Behavior

The baseline component currently requires a pseudo-stack to output the name and object-id of an Administrators group that should be granted access to the Azure Keyvault. The baseline component then grants access to the keyvault for that group. Without it, the keyvault cannot be double-checked through the Azure Portal as the user will not have permissions.

## Possible Solution

Update the existing Prologue template pass for the baseline to include a new administrators account and output the pseudo-stack prior to the template pass.

## Context

Prevented me from deploying the baseline component into a fresh environment.

### LINKED ISSUE #179: Baseline deployment should fail if user can't encrypt key
## Expected Behaviour
An exception is raised and deployment cancelled when the user running a template deployment doesn't have an access policy allowing them to access the Key Vault 

## Current Behaviour
The command line error messages is displayed in the console but the deployment continues

## Possible Solution
When an exception is raised in the encryption steps, exit and cancel the  deploy

## Steps to Reproduce (for bugs)
1. run a deployment of the baseline component without permissions assinged on the key vault

## Context
False positive result for the deployment

### PR DESCRIPTION
**Pull request by ken9ross**
Fixes #120 
Fixes #179
Fixes https://github.com/hamlet-io/engine-plugin-azure/issues/176 by superceding the value and setting a default
Change Azure keyVault to use azure:AdministratorGroups

## Description
Shift to name-spaced component attributes from setting. Update the existing Prologue template pass for the baseline to include a new administrators account and output the pseudo-stack prior to the template pass.

## Motivation and Context
Move to new standard approach

## How Has This Been Tested?
Local template generation and subsequent deployment

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Followup Actions
- [x] Any existing Azure products will need to change to use azure:AdministratorGroups for keyVault
- [ ] None

## Checklist:

- [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] None of the above.

### CODE REVIEW

**Comment by RossMurr4y** [comment]
Looks like the `baseline` defined in the Masterdata needs to have the new mandatory attribute added.

File: azuretest/inputsources/shared/blueprint.ftl [hunk 1]
```diff
@@ -67,6 +67,17 @@
...
+            },
+            "Tiers" : {
+                "mgmt" : {
+                    "Components" : {
+                        "baseline" : {
+                            "baseline" : {
+                                "azure:AdministratorGroups" : [ "1234567890-1234567890-1234567890-1234567890" ]
+                            }
+                        }
+                    }
```
Comment (roleyfoley) [inline azuretest/inputsources/shared/blueprint.ftl:79]: Should this be in the mock input rather than the shared one? This would be the default value whenever someone doesn't set the value

**Review by roleyfoley — CHANGES_REQUESTED** [review]

File: azuretest/inputsources/shared/blueprint.ftl (hunk 1 above)
Comment (ken9ross) [reply azuretest/inputsources/shared/blueprint.ftl:79]: I thought azuretest was all mock. Have I misunderstood what azuretest is?

Comment (roleyfoley) [reply azuretest/inputsources/shared/blueprint.ftl:79]: Ah right missed that sorry, yeah its ok to go here 
The other place that might be better is in the baseline testing module, all the other components will still use it and keeps the baseline configuration in a centralised point

Comment (ken9ross) [reply azuretest/inputsources/shared/blueprint.ftl]: Will do

**Comment by roleyfoley** [comment]
@ken9ross I think you might need to rebase this PR seems to contain changes from https://github.com/hamlet-io/engine-plugin-azure/pull/218

**Comment by RossMurr4y** [comment]
@ken9ross if this ticket addresses #179 then using the closing-syntax here will close that ticket when this is merged. Presently the reference is the other way around, which won't have the same effect

**Comment by ken9ross** [comment]
@roleyfoley I think I've cleaned it up, but I don't know how to confirm it. Any advice?

**Comment by roleyfoley** [comment]
The files changed tab shows that your changes are now only yours
The commit history on the PR should only include commits that you've made, it shouldn't include others.

File: azure/components/baseline/setup.ftl [hunk 2]
```diff
@@ -260,9 +261,15 @@
...
-                  "  if [[ ! $(az_check_secret" + " " +
+                  "  AZ_CHK_SECRET=$(az_check_secret" + " " +
                       "\"" + vmKeyVaultName + "\" " +
-                      "\"" + vmKeyPairName + "PublicKey" + "\") " +
+                      "\"" + vmKeyPairName + "PublicKey" + "\")",
+                  "  echo $\{AZ_CHK_SECRET}",
+                  "  if [[ $\{AZ_CHK_SECRET} " +
+                      "=~ \"does not have secrets get permission on key vault\" ]]; then",
+                  "    fatal \"The deployment user is not a member of the specified keyVault admin group\"",
+                  "  else",
```
Comment (roleyfoley) [inline azure/components/baseline/setup.ftl:271]: Might be better to break these into two if statements 
so the fatal section can exit with a non 0 return code to stop any futher scripts running 

Then you don't need a nested if statement. 

If that doesn't work the second if should be indented rather than inline with the initial if

Comment (roleyfoley) [reply azure/components/baseline/setup.ftl:271]: The other alternative is elif, which would be ok to use in this case as its a block related to a single task

File: azure/components/baseline/setup.ftl [hunk 3]
```diff
@@ -320,6 +328,33 @@
...
+        [#local adminId = formatId(keyvaultId,adminGrp)]
+        [@addToDefaultBashScriptOutput
+          content=
+          [
+            "case $\{DEPLOYMENT_OPERATION} in",
+            "  create|update)",
+            "    ADMINGRP=$(az role definition list --name "+adminGrp+")",
+            "    if [[ $\{#ADMINGRP[@]} -eq 0 ]] ; then",
+            "      fatal \"Azure Administrator role does not exist: "+adminGrp+"\"",
+            "    else"
```
Comment (roleyfoley) [inline azure/components/baseline/setup.ftl:342]: That should end with `fi` instead of if and also we should exit after the fatal if none of the other steps can be run

File: azure/components/baseline/setup.ftl [hunk 4]
```diff
@@ -320,6 +328,33 @@
...
+            "AdministratorGroups",
+            { adminId : adminGrp },
+            "admingrp"
+          ) +
+          [
+            "    fi",
+            "       ;;",
+            "       esac"
+          ]
+        /]
```
Comment (roleyfoley) [inline azure/components/baseline/setup.ftl:354]: This is quite hard to follow In line with the comment above I would remove it

File: azure/inputsources/shared/masterdata.ftl [hunk 5]
```diff
@@ -853,7 +853,8 @@
                                 "oai": {
                                     "Engine": "oai"
                                 }
-                            }
+                            },
+                            "azure:AdministratorGroups": []
```
Comment (roleyfoley) [inline azure/inputsources/shared/masterdata.ftl:857]: Not required, the default value is set on the attribute anyway

**Review by roleyfoley — CHANGES_REQUESTED** [review]

File: azure/components/baseline/setup.ftl (hunk 2 above)
Comment (ken9ross) [reply azure/components/baseline/setup.ftl:271]: If you aren't a member of the admin group it doesn't make sense to try and do the private key having failed on the public key. I've added the indenting

File: azure/components/baseline/setup.ftl (hunk 3 above)
Comment (ken9ross) [reply azure/components/baseline/setup.ftl:342]: The `fi` is at L350 to make the pseudoArmStackOutputScript() conditional, or are you talking about elsewhere?

File: azure/inputsources/shared/masterdata.ftl (hunk 5 above)
Comment (ken9ross) [reply azure/inputsources/shared/masterdata.ftl:857]: The default value causes an error to be raised ("no administrator group"). We're trying to find a better default value so the default "just works", so this is part default value, part placeholder for a better default value. The default on the attribute is to avoid any crashes (as distinct from deliberate error reporting)

File: azure/components/baseline/setup.ftl (hunk 2 above)
Comment (roleyfoley) [reply azure/components/baseline/setup.ftl:271]: If that configuration isn't setup properly is there point in continuing with the rest of the script? 
the `fatal` function is a fatal log message and doesn't produce and non 0 exit code, so based on the current script it would log a fatal error but complete successfully. 

So it might be better to handle any fatal errors as indivdual steps and exit with a non 0 exit code then process the happy results once we have got past the requirements

File: azure/inputsources/shared/masterdata.ftl (hunk 5 above)
Comment (roleyfoley) [reply azure/inputsources/shared/masterdata.ftl:857]: Where are you seeing an error? It might be better to check for content in the code which uses it and if there isn't content skip the step. This would then allow the validation failure to get out which is probably where the issue is at the moment.

**Review by roleyfoley — APPROVED** [review]
